### PR TITLE
convert Microsoft.AVS 2023-03-01 to TypeSpec

### PR DESCRIPTION
**Pull request by cataggar**

### CODE REVIEW

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 1]
```diff
@@ -0,0 +1,1362 @@
...
+using TypeSpec.Rest;
+using TypeSpec.Http;
+using Azure.ResourceManager;
+
+namespace Microsoft.AVS;
+
+interface Operations extends Azure.ResourceManager.Operations {}
+
+interface Locations {
+  @doc("Return trial status for subscription by region")
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp:15]: You will likely want to use either the checkGlobal or checkLocal name availability operation templates: https://github.com/Azure/typespec-azure/blob/main/packages/typespec-azure-resource-manager/lib/operations.tsp#L435-L471

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 2]
```diff
@@ -0,0 +1,1362 @@
...
+    {
+      ...SubscriptionIdParameter;
+      ...ResourceGroupParameter;
+      ...PrivateCloudNameParameter;
+    },
+    AdminCredentials
+  >;
+}
+
+interface Clusters {
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp:71]: You should just add Cluster as a child resource, by setting `@parentResource` for cluster, as described here: https://azure.github.io/typespec-azure/docs/getstarted/azure-resource-manager/step03

Then you can just use the standard operation templates.

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 3]
```diff
@@ -0,0 +1,1362 @@
...
+      ...SubscriptionIdParameter;
+      ...ResourceGroupParameter;
+      ...PrivateCloudNameParameter;
+      ...ClusterNameParameter;
+    },
+    ClusterZoneList
+  >;
+}
+
+interface Datastores {
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp:87]: same comment as above, DataStore should just be a child resource of Cluster

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 4]
```diff
@@ -0,0 +1,1362 @@
...
+      ...ResourceGroupParameter;
+      ...PrivateCloudNameParameter;
+      ...ClusterNameParameter;
+      ...DatastoreNameParameter;
+    },
+    void
+  >;
+}
+
+interface HcxEnterpriseSites {
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp:100]: hexEnterpriseSItes should be a child resource of privateClouds, as above

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 5]
```diff
@@ -0,0 +1,1362 @@
...
+      ...SubscriptionIdParameter;
+      ...ResourceGroupParameter;
+      ...PrivateCloudNameParameter;
+      ...HcxEnterpriseSiteNameParameter;
+    },
+    void
+  >;
+}
+
+interface Authorizations {
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp:113]: model as a child resource

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 6]
```diff
@@ -0,0 +1,1362 @@
...
+      ...SubscriptionIdParameter;
+      ...ResourceGroupParameter;
+      ...PrivateCloudNameParameter;
+      ...AuthorizationNameParameter;
+    },
+    void
+  >;
+}
+
+interface GlobalReachConnections {
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp:126]: model as a chile resource

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 7]
```diff
@@ -0,0 +1,1362 @@
...
+      ...SubscriptionIdParameter;
+      ...ResourceGroupParameter;
+      ...PrivateCloudNameParameter;
+      ...GlobalReachConnectionNameParameter;
+    },
+    void
+  >;
+}
+
+interface WorkloadNetworks {
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp:139]: model as a child resource

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 8]
```diff
@@ -0,0 +1,1362 @@
...
+      ...SubscriptionIdParameter;
+      ...ResourceGroupParameter;
+      ...PrivateCloudNameParameter;
+      ...SegmentIdParameter;
+    },
+    void
+  >;
+
+  @doc("List dhcp in a private cloud workload network.")
+  @route("/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AVS/privateClouds/{privateCloudName}/workloadNetworks/default/dhcpConfigurations")
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp]: workloadNetworks should be modeld as a singleton resource: https://azure.github.io/typespec-azure/docs/howtos/ARM/resource-type#singleton-resource

Comment (markcowl) [reply specification/vmware/Microsoft.AVS/routes.tsp]: And dhcpConfiguration should be a child of this resource

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 9]
```diff
@@ -0,0 +1,1362 @@
...
+    {
+      ...SubscriptionIdParameter;
+      ...ResourceGroupParameter;
+      ...PrivateCloudNameParameter;
+    },
+    WorkloadNetworkGatewayList
+  >;
+
+  @doc("Get a gateway by id in a private cloud workload network.")
+  @route("/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AVS/privateClouds/{privateCloudName}/workloadNetworks/default/gateways/{gatewayId}")
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp]: child resource

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 10]
```diff
@@ -0,0 +1,1362 @@
...
+    {
+      ...SubscriptionIdParameter;
+      ...ResourceGroupParameter;
+      ...PrivateCloudNameParameter;
+    },
+    WorkloadNetworkPortMirroringList
+  >;
+
+  @doc("Get a port mirroring profile by id in a private cloud workload network.")
+  @route("/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AVS/privateClouds/{privateCloudName}/workloadNetworks/default/portMirroringProfiles/{portMirroringId}")
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp]: child resource

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 11]
```diff
@@ -0,0 +1,1362 @@
...
+      ...PrivateCloudNameParameter;
+    },
+    void
+  >;
+
+  #suppress "@azure-tools/typespec-azure-core/casing-style" "Can not change existing name."
+  @doc("List of vm groups in a private cloud workload network.")
+  @route("/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AVS/privateClouds/{privateCloudName}/workloadNetworks/default/vmGroups")
+  @get
+  listVMGroups is Azure.Core.Foundations.Operation<
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp]: child resource

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 12]
```diff
@@ -0,0 +1,1362 @@
...
+      ...ResourceGroupParameter;
+      ...PrivateCloudNameParameter;
+    },
+    WorkloadNetworkVirtualMachinesList
+  >;
+
+  @doc("Get a virtual machine by id in a private cloud workload network.")
+  @route("/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AVS/privateClouds/{privateCloudName}/workloadNetworks/default/virtualMachines/{virtualMachineId}")
+  @get
+  getVirtualMachine is Azure.Core.Foundations.Operation<
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp]: child resource

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 13]
```diff
@@ -0,0 +1,1362 @@
...
+      ...PrivateCloudNameParameter;
+      ...VirtualMachineIdParameter;
+    },
+    WorkloadNetworkVirtualMachine
+  >;
+
+  @doc("List of DNS services in a private cloud workload network.")
+  @route("/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AVS/privateClouds/{privateCloudName}/workloadNetworks/default/dnsServices")
+  @get
+  listDnsServices is Azure.Core.Foundations.Operation<
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp]: child resource

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 14]
```diff
@@ -0,0 +1,1362 @@
...
+      ...DnsServiceIdParameter;
+      ...PrivateCloudNameParameter;
+    },
+    void
+  >;
+
+  @doc("List of DNS zones in a private cloud workload network.")
+  @route("/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AVS/privateClouds/{privateCloudName}/workloadNetworks/default/dnsZones")
+  @get
+  listDnsZones is Azure.Core.Foundations.Operation<
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp]: child

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 15]
```diff
@@ -0,0 +1,1362 @@
...
+      ...PrivateCloudNameParameter;
+    },
+    void
+  >;
+
+  #suppress "@azure-tools/typespec-azure-core/casing-style" "Can not change existing name."
+  @doc("List of Public IP Blocks in a private cloud workload network.")
+  @route("/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AVS/privateClouds/{privateCloudName}/workloadNetworks/default/publicIPs")
+  @get
+  listPublicIPs is Azure.Core.Foundations.Operation<
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp]: child resource

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 16]
```diff
@@ -0,0 +1,1362 @@
...
+      ...SubscriptionIdParameter;
+      ...ResourceGroupParameter;
+      ...PublicIpIdParameter;
+      ...PrivateCloudNameParameter;
+    },
+    void
+  >;
+}
+
+interface CloudLinks {
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp:343]: child resource

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 17]
```diff
@@ -0,0 +1,1362 @@
...
+      ...SubscriptionIdParameter;
+      ...ResourceGroupParameter;
+      ...PrivateCloudNameParameter;
+      ...CloudLinkNameParameter;
+    },
+    void
+  >;
+}
+
+interface Addons {
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp:356]: child resource

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 18]
```diff
@@ -0,0 +1,1362 @@
...
+      ...SubscriptionIdParameter;
+      ...ResourceGroupParameter;
+      ...PrivateCloudNameParameter;
+      ...AddonNameParameter;
+    },
+    void
+  >;
+}
+
+interface VirtualMachines {
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp:369]: child resource

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 19]
```diff
@@ -0,0 +1,1362 @@
...
+      ...ResourceGroupParameter;
+      ...PrivateCloudNameParameter;
+      ...ClusterNameParameter;
+      ...PlacementPolicyNameParameter;
+    },
+    PlacementPolicy
+  >;
+
+  @doc("Create or update a placement policy in a private cloud cluster")
+  @route("/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AVS/privateClouds/{privateCloudName}/clusters/{clusterName}/placementPolicies/{placementPolicyName}")
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp]: child resource

File: specification/vmware/Microsoft.AVS/routes.tsp [hunk 20]
```diff
@@ -0,0 +1,1362 @@
...
+
+      @doc("A placement policy in the private cloud cluster")
+      @body
+      placementPolicy: PlacementPolicy;
+    },
+    PlacementPolicy
+  >;
+
+  @doc("Update a placement policy in a private cloud cluster")
+  @route("/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.AVS/privateClouds/{privateCloudName}/clusters/{clusterName}/placementPolicies/{placementPolicyName}")
```
Comment (markcowl) [inline specification/vmware/Microsoft.AVS/routes.tsp]: In geenral, you should never have `@route` in an ARM typespec.  If your resources are modeled as parent/children, you should be able to use the standard operation templates

**Review by markcowl — CHANGES_REQUESTED** [review]
A small bit of remodeling would vastly simplify this spec, simply by using child resources and singleton resources, and by using the existing checkNamevailability operation templates

**Comment by cataggar** [comment]
Replaced by https://github.com/Azure/azure-rest-api-specs-pr/pull/15631

**Pull request closed without merging**